app/gui2/src/util/ast/__tests__/aliasAnalysis.test.ts

```diff
@@ -0,0 +1,319 @@
+      }
+    }
+
+    // Unresolved symbols
+    for (const unresolvedSymbol of this.unresolvedSymbols) {
+      assert(
+        analyzer.unresolvedSymbols.has(unresolvedSymbol),
+        `Expected unresolved symbol usage at [${unresolvedSymbol}] not observed.`,
+      )
+      console.log(`Found expected unresolved symbol usage at [${unresolvedSymbol}]`)
+    }
+
+    console.log(analyzer.unresolvedSymbols)
+    return analyzer
+  }
+
+  static parseAndRun(annotatedCode: string) {
+    const testCase = TestCase.parse(annotatedCode)
+    return testCase.run()
+  }
+}
+
+if (import.meta.vitest) {
+  const { test } = import.meta.vitest
+
+  test('Annotations parsing', () => {
+    const annotatedCode = `main =
+    «1,x» = 1
+    «2,y» = 2
+    z = «3,x» -> »3,x« + »2,y«
+    u = »1,x« + »2,y«`
+
+    const expectedUnannotatedCode = `main =
+    x = 1
+    y = 2
+    z = x -> x + y
+    u = x + y`
+
+    const { unannotatedCode, annotations } = parseAnnotations(annotatedCode)
+    assert(unannotatedCode === expectedUnannotatedCode)
+    assert(annotations.size === 7)
+    const validateAnnotation = (
+      range: ContentRange,
+      kind: AnnotationType,
+      prefix: number,
+      identifier: string,
+    ) => {
+      try {
+        const a = annotations.get(range)
+        assertDefined(a, `Annotation is not defined.`)
+        assertEqual(a.kind, kind, 'Invalid annotation kind.')
+        assertEqual(a.id, prefix, 'Invalid annotation prefix.')
+        assertEqual(
+          unannotatedCode.substring(range[0], range[1]),
+          identifier,
+          'Invalid annotation identifier.',
+        )
+      } catch (e) {
+        const message = `Invalid annotation at [${range}]: ${e}`
+        throw new Error(message)
+      }
+    }
+
+    validateAnnotation([11, 12], AnnotationType.Binding, 1, 'x')
+    validateAnnotation([21, 22], AnnotationType.Binding, 2, 'y')
+    validateAnnotation([35, 36], AnnotationType.Binding, 3, 'x')
+    validateAnnotation([40, 41], AnnotationType.Usage, 3, 'x')
+    validateAnnotation([44, 45], AnnotationType.Usage, 2, 'y')
+    validateAnnotation([54, 55], AnnotationType.Usage, 1, 'x')
+    validateAnnotation([58, 59], AnnotationType.Usage, 2, 'y')
+  })
+
+  test('Plain dependency', () => {
+    const code = 'main =\n    «1,x» = 1\n    »1,x«'
+    TestCase.parseAndRun(code)
+  })
+
+  test('Unresolved dependency', () => {
+    const code = 'main =\n    «1,x» = 1\n    »2,y«'
+    TestCase.parseAndRun(code)
+  })
+
+  test('Local variable shadowed by the lambda argument', () => {
+    const code = `main =
+    «1,x» = 1
+    «2,y» = 2
+    z = «3,x» -> »3,x« + »2,y«
+    u = »1,x« + »2,y«`
+
+    TestCase.parseAndRun(code)
+  })
+
+  test('Plain symbol usage', () => {
+    const code = '»1,x«'
+    TestCase.parseAndRun(code)
+  })
+
+  test('Accessors', () => {
+    const code = `main =
+    «1,x» = 1
+    »1,x«.x.y »2,arg«`
+    TestCase.parseAndRun(code)
+  })
+
+  test('Function argument', () => {
```

Comment:
Consider using [`test.each`](https://vitest.dev/api/#test-each) vitest API instead of individual test functions that do exactly the same thing with slightly different input.

Reply:
I've tried to set this up but `each` does not play nicely with the IDE. Using it breaks the possibility to run or debug test from the text editor. 
Moreover, it makes it impossible to run/debug the test individually from the IDE, which is a capability that I heavily rely on.